### PR TITLE
Rust/His: Speed up generalized pattern checks

### PR DESCRIPTION
**Pull request by beaubelgrave**
The existing generalized pattern check code simply linearly scans all definitions for matching packed words. This works well with a few definitions, but as more are added the performance suffers.

Create 32 lanes for UTF8 and UTF16 that split up the packed words. Use the accumulator value to find the right lane to check, then linear scan those. This makes the linear scan faster as UTF8 and UTF16 are isolated to "like" patterns.

### CODE REVIEW

File: src/Microsoft.Security.Utilities.Core/IdentifiableScan.cs [hunk 1]
```diff
@@ -123,7 +125,19 @@ public void Start()
     {
         if (this.scan == IntPtr.Zero)
         {
-            this.scan = identifiable_scan_create();
+            var builder = new StringBuilder();
+
+            foreach (var id in this.idToLengthMap.Keys)
```
Comment (michaelcfanning) [inline src/Microsoft.Security.Utilities.Core/IdentifiableScan.cs:134]: >var [](http://example.com/codeflow?start=21&length=3)

we'd typically be explicit about type here,

File: src/Microsoft.Security.Utilities.Core/IdentifiableScan.cs [hunk 2]
```diff
@@ -123,7 +125,19 @@ public void Start()
...
+            {
+                builder.Append(id);
+		builder.Append(';');
```
Comment (michaelcfanning) [inline src/Microsoft.Security.Utilities.Core/IdentifiableScan.cs]: >builder [](http://example.com/codeflow?start=2&length=7)

NIT, indentation is a little off.

Comment (michaelcfanning) [inline src/Microsoft.Security.Utilities.Core/IdentifiableScan.cs]: >builder [](http://example.com/codeflow?start=2&length=7)

nit, our convention for dropping lines on parameters is for the first arg to remain on the same line as the call, and args to left-align underneath that location.

an old Windows C++ convention i came to like. :)

Comment (beaubelgrave) [reply src/Microsoft.Security.Utilities.Core/IdentifiableScan.cs]: Fixed to just a single line, since only 2 args.

Comment (beaubelgrave) [reply src/Microsoft.Security.Utilities.Core/IdentifiableScan.cs]: Fixed.

File: src/Microsoft.Security.Utilities.Core/IdentifiableScan.cs (hunk 1 above)
Comment (beaubelgrave) [reply src/Microsoft.Security.Utilities.Core/IdentifiableScan.cs:134]: Moved to ThreadStatic as suggested offline.

File: src/Microsoft.Security.Utilities.Core/IdentifiableScan.cs [hunk 3]
```diff
@@ -123,7 +128,18 @@ public void Start()
     {
         if (this.scan == IntPtr.Zero)
         {
-            this.scan = identifiable_scan_create();
+            stringBuilder ??= new StringBuilder();
```
Comment (michaelcfanning) [inline src/Microsoft.Security.Utilities.Core/IdentifiableScan.cs:131]: >StringBuilder [](http://example.com/codeflow?start=34&length=13)

this is exactly what i asked for so why would i complain? but on second reflection, probably we could have just had a per-instance item here. 

but let's leave it, it's not worth going back. thanks for acting on the notes.

**Review by michaelcfanning — APPROVED** [review]
:shipit:

**Pull request merged**